Python/chapter02/2.2 - Return Kth to Last/miguel_2.2_soln.py

```diff
@@ -0,0 +1,163 @@
+"""
+Python version 3.7.0
+2.2 - Return Kth to Last
+Implement an algorithm to find the kth
+to last element of a singly linked list
+Example:
+    input: ll = a -> b -> c -> d -> e -> f
+            k = 3
+    output: d because d is 3rd to last
+"""
+import unittest
+
+
+class Node:
+    def __init__(self, d: int):
+        self.data = d
+        self.next = None
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __str__(self):
+        return '<Node Value: {}>'.format(self.data)
+
+    def __eq__(self, other: object):
+        if not isinstance(other, Node):
+            return NotImplemented
+        return self.data == other.data
+
+
+class LinkedList:
+    def __init__(self, *numbers: int):
+        self.head = None
+        self.tail = None
+        for num in numbers:
+            self.append_to_tail(num)
+
+    def append_to_tail(self, d: int) -> None:
+        if self.head is None:
+            self.head = Node(d)
+            self.tail = self.head
+            return
+        end = Node(d)
+        self.tail.next = end
+        self.tail = end
+
+    def append_to_head(self, d: int) -> None:
+        new_head = Node(d)
+        new_head.next = self.head
+        self.head = new_head
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __str__(self):
+        if self.head is None:
+            return '<empty>'
+        ll = []
+        n = self.head
+        while n.next is not None:
+            ll.append('{} -> '.format(n.data))
+            n = n.next
+        ll.append(str(n.data))
+        return ''.join(ll)
+
+    def __eq__(self, other: object):
+        if not isinstance(other, LinkedList):
+            return NotImplemented
+        a = self.head
+        b = other.head
+        while a is not None and b is not None:
+            if a.data != b.data:
+                return False
+            # otherwise, advance both pointers
+            a = a.next
+            b = b.next
+        return a is None and b is None
+
+
+def kth_to_last(ll: LinkedList, k: int) -> Node:
+    """
+    kth_to_last will return the kth to last node
+    from the input linked list.
+    Going to reverse the linked list and then
+    count k steps.
+    Runtime: O(N)
+    Space Complexity: O(N)
```

Comment:
is there a O(1) answer? is there an answer where you don't explicitly allocate any memory and instead let the runtime handle that? (you might think that is cheating because it would still be O(n) space complexity, but it "feels" O(1)).

Reply:
I'm thinking that I can loop through the linked list to get the size, then traverse to the size - kth node. (starting i at 1).  No extra linked list created.

Maybe it would be better to keep track of the linked list's size as a member variable?

Reply:
> I'm thinking that I can loop through the linked list to get the size, then traverse to the size - kth node. (starting i at 1). No extra linked list created.

that works.

> Maybe it would be better to keep track of the linked list's size as a member variable?

that's definitely an option. C++ "recently" changed the standard to actually specify that: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2009/n2923.pdf